[V2V] Refactor ConversionHost to use AuthenticationMixin

djberg96 opened the pull request:
Currently there's a lot of hand-crafted ssh handling using the MiqSshUtil library baked into the ConversionHost model that is tied to specific providers. I thought we should leverage the AuthenticationMixin to take advantage of its features.

Specifically, this would give use a direct association between conversion hosts and authentications. This would be beneficial for certain providers, e.g. openstack, where we are currently using credentials set at the provider level. This way, we can set it on a per-resource level for all provider types by setting the authentication on the conversion host instance itself.

This will also be necessary for the UI where users are allowed to upload their own ssh keys.

djberg96 commented:
@fdupont-redhat @agrare Just looking for some feedback here to see if it makes sense to leverage the AuthenticationMixin instead of the current code, while seems a bit clumsy to me.

agrare commented:
Hm I think we were relying on the fact that the underlying resource had the authentication not the conversion host?  And a Vm and a Host both include AuthenticationMixin afaik

fdupont-redhat commented:
We don't exactly rely on the resource authentications. In the case of OpenStack for example, the authentication is associated with the provider, forcing us to have a single key for all conversion hosts of the provider. It was implemented this way to simplify the UX, as it only required to expose the ssh_keypair authentication for the provider. It might be a good idea to revisit this topic with advanced UI for conversion host.

agrare commented on app/models/conversion_host.rb:
```diff
@@ -21,31 +22,81 @@ class ConversionHost < ApplicationRecord
+  #
+  # This method is necessary to comply with AuthenticationMixin interface.
+  #
+  def verify_credentials(_auth_type = nil, _options = {})
```
Hm we already have a common `verify_credentials_with_ssh` method for hosts [`Host#verify_credentials_with_ssh`](https://github.com/ManageIQ/manageiq/blob/master/app/models/host.rb#L769-L789), is that method something that you could use?  We could move it into a common place like the `AuthenticationMixin`

djberg96 replied:
@agrare In theory yes, though I think there are some implementation details in there that might not be right for conversion hosts, e.g. I'm not sure we're pinned to Linux vms.

agrare replied:
I'm not sure that check was for anything else than failing fast for hosts that we don't expect to even have the ssh daemon?  We could refactor that method if we need.

agrare replied:
Also, are we expecting people to deploy conversion hosts on BSD frequently? :laughing:

djberg96 replied:
I dunno, maybe even Solaris? Didn't want to make any assumptions. :)

djberg96 replied:
@agrare I think I'd like to hold off on that refactoring, in part because I think MiqSshUtil is overworked for just testing a connection. Looks to me like it's meant for cases where we actually care about data, and here we don't.

agrare commented on app/models/conversion_host.rb:
```diff
@@ -21,31 +26,97 @@ class ConversionHost < ApplicationRecord
+      require 'net/ssh'
+      host = hostname || ipaddress
+
+      authentications.each do |auth|
```
Typically the auth_type is used to pick a single authentication to use rather than looping through all authentications.

If you're looking for username/password you'd use `userid, password = auth_user_pwd(auth_type)`

djberg96 replied:
I don't think auth_type would ever be set, though.

djberg96 replied:
Alright, I updated it to use the `auth_type`, but it required modifying the authentication mixin because it wasn't looking for private keys as a possible result.

agrare commented on app/models/conversion_host.rb:
```diff
@@ -21,31 +26,97 @@ class ConversionHost < ApplicationRecord
+      host = hostname || ipaddress
+
+      authentications.each do |auth|
+        user = auth.userid || ENV['USER'] || Etc.getlogin
```
Do we need the other options?  I'd much rather require people enter a username instead of trying to pull a username from a few different places.

djberg96 replied:
Perhaps I was trying to be too helpful here.

agrare commented on app/models/conversion_host.rb:
```diff
@@ -21,31 +26,97 @@ class ConversionHost < ApplicationRecord
+          ssh_options[:auth_methods] = %w[password]
+        end
+
+        if auth.auth_key
```
So if you need more than userid+password you can use `auth = authentication_type(auth_type)` instead of what I said above.

djberg96 replied:
Since this is an internal check, I don't think `auth_type` would ever be set, and I had a hard time finding examples in other code where it wasn't nil.

In practice, I doubt there would ever be more than one associated authentication, but I figured if more than one ever was set, they should all be checked.

agrare replied:
> Since this is an internal check, I don't think auth_type would ever be set, and I had a hard time finding examples in other code where it wasn't nil.

I don't think I've ever seen it not be set...

```
=> [["ManageIQ::Providers::Vmware::InfraManager dev-vc65", "default"], ["ManageIQ::Providers::Vmware::InfraManager::HostEsx dell-r430-15.cloudforms.lab.eng.rdu2.redhat.com", "default"], ["ManageIQ::Providers::Vmware::InfraManager::HostEsx dell-r430-16.cloudforms.lab.eng.rdu2.redhat.com", "default"], ["ManageIQ::Providers::Vmware::InfraManager::HostEsx dell-r430-17.cloudforms.lab.eng.rdu2.redhat.com", "default"]]
```

> In practice, I doubt there would ever be more than one associated authentication, but I figured if more than one ever was set, they should all be checked.

This is probably true so the authtype should be `"default"`

djberg96 replied:
Where/how does this get selected? Is this an external option coming from the UI? I guess I'm not sure how it fits here, since we're relying on the properties of the associated authentications to pass the proper options. Allowing an auth_type wouldn't be sensical to me. Perhaps I'm missing some context?

djberg96 replied:
Updated to use auth_type first, which will return a result if not nil.

agrare commented on app/models/conversion_host.rb:
```diff
@@ -21,31 +26,97 @@ class ConversionHost < ApplicationRecord
+  # command on the connection. The exact nature of the connection will depend on the
+  # underlying provider.
+  #
   def check_ssh_connection
```
Now is the point of this method to see if we can ssh without username/password aka someone set the appliance's root ssh public key to the conversion host's authorized_keys ?

djberg96 replied:
Near as I could tell the purpose was just to check that it could make a connection, whether it be via authkey or userid/password or whatever.

agrare replied:
I wonder if we can delete it then?  @fdupont-redhat do you remember what this was for?

fdupont-redhat replied:
The idea was to validate the connection by running a command and not stopping at negotiation... If this validation is provided by AuthenticationMixin, we can remove it.

djberg96 replied:
This method would still be used in cases where there wasn't an associated authentication, so we still need it for now. In that case it would default to the resource authentication (redhat) or the provider (openstack).

agrare replied:
Ohh I see connect_ssh is actually just username/password from the related resource, I thought that was the MiqSshUtil method and was relying on passwordless ssh

djberg96 commented:
@agrare Made a couple changes. First, it will now use `auth_type` if present, using the mixed-in `authentication_type` method. To make that work properly however, I had to modify the AuthenticationMixin by adding a `authentication_private_keys` method (surprisingly not already present) and then modifying `available_authentications` to include them.

Second, I came around to the idea that iterating over every authentication was not necessary as there should only ever be one associated authentication. Plus, it simplified the code.

I've updated the specs, too. Please let me know if there are any other issues and/or if you want me to squash these commits.

mturley commented:
This one should also be `hammer/yes` and listed in https://bugzilla.redhat.com/show_bug.cgi?id=1622728

fdupont-redhat commented:
@miq-bot add_label transformation, enhancement, hammer/yes

agrare commented on app/models/conversion_host.rb:
```diff
@@ -124,28 +208,61 @@ def resource_supports_conversion_host
+    authentication = AuthUseridPassword.find_by(:resource => resource)
+
+    if authentication.nil? && resource.authentication_userid && resource.authentication_password
+      authentication = AuthUseridPassword.new(
```
Why is a method called `find_` creating new authentications?  Also this won't create the record without .save!

djberg96 replied:
Right, I'm deliberately not saving it. The goal is a common object that I can then call methods on, instead of dealing with different return types depending on the provider, which is clumsy.

agrare replied:
...interesting could we just return the `resource.authentication_type()`

agrare replied:
This is still an issue, why not just return the resource authentication instead of creating a new one?

djberg96 replied:
Mainly this was done for openstack, where we want an authentication independent of the provider. But even for other providers, I felt it was more flexible to have a separate authentication.

@fdupont-redhat Am I incorrect?

agrare replied:
I don't mean why have a separate auth for a conversation host in general, I mean here where you're essentially just using the resources authentication just return that one like you do with openstack below instead if creating a temporary object here

djberg96 replied:
Ok, updated. I used the infra host.rb file in the ovirt provider code as a guideline.

agrare commented on app/models/conversion_host.rb:
```diff
@@ -26,31 +31,99 @@ class ConversionHost < ApplicationRecord
+      host = hostname || ipaddress
+
+      auth = authentication_type(auth_type) || authentications.first
+      user = auth.userid || ENV['USER']
```
Not a fan of the `ENV['USER']`, it _might_ help with development but no one on an appliance is going to do this

djberg96 replied:
Yeah, it's there for development mainly. Unless you explicitly object, I'd like to leave it.

agrare replied:
We should remove it, it is simple enough to set a userid on the auth record for development anyway

djberg96 replied:
Ok, done.

agrare commented on app/models/conversion_host.rb:
```diff
@@ -26,31 +31,99 @@ class ConversionHost < ApplicationRecord
+      ssh_options = { :timeout => 10, :logger => $log, :verbose => :error }
+
+      case auth
+        when AuthUseridPassword
```
This should be aligned with the `case`

djberg96 replied:
ok, done.

agrare commented on app/models/conversion_host.rb:
```diff
@@ -26,31 +31,99 @@ class ConversionHost < ApplicationRecord
     connect_ssh { |ssu| ssu.shell_exec('uname -a') }
     true
-  rescue => e
+  rescue StandardError
```
I don't agree with rubocop on this one, also would rather catch the exception and log it

djberg96 replied:
MiqSshUtil will already do some logging.

agrare replied:
Okay

agrare commented on app/models/conversion_host.rb:
```diff
@@ -124,28 +208,61 @@ def resource_supports_conversion_host
+  # finding credentials.
+  #
+  def find_credentials(msg = nil)
+    authentication = send("find_resource_credentials_#{resource.ext_management_system.emstype}")
```
Can we just do this in a provider subclass now?

djberg96 replied:
That's the plan eventually, want to leave it as a TODO for now.

djberg96 replied:
Actually, this was simplified so that it's no longer provider dependent. Though, of course, it could be redefined if ever needed.

agrare replied:
:+1:

agrare commented on app/models/mixins/authentication_mixin.rb:
```diff
@@ -70,6 +70,10 @@ def authentication_key_pairs
   end
 
+  def authentication_private_keys
+    authentications.select { |a| a.kind_of?(AuthPrivateKey) }
```
I think that an Openstack AuthKeyPair is the same as an AuthPrivateKey so we might be able to just update https://github.com/ManageIQ/manageiq/pull/18309/files#diff-f9cace1fe4fdcec66f962b23f2110fd6R70

djberg96 replied:
You mean change `ManageIQ::Providers::Openstack::InfraManager::AuthKeyPair` to just `AuthKeyPair`? Otherwise I'm not sure what you mean. Also, I thought about changing that line, but was worried about introducing some sort of regression.

agrare replied:
Yes exactly, rather than add a completely new type I'd rather combine the two.

djberg96 replied:
Actually, I just noticed something. It's looking for INFRA AuthKeyPair's. I was thinking Cloud this whole time, duh. Given that there's no AuthKeyPair base class, and the subclass is just an empty subclass, I just removed `authentication_key_pairs` completely as `authentication_private_keys` will pick them up. I can only speculate as to what the plans were for that class, but it looks unnecessary.

Or maybe that's what you meant this whole time and we are now in agreement. Anyway, updated. Please let me know if I've misinterpreted what you want.

agrare replied:
> Or maybe that's what you meant this whole time and we are now in agreement. Anyway, updated. Please let me know if I've misinterpreted what you want.

Exactly :)  I think it'd be safer to use the old method but just change it to look for `AuthPrivateKey` since the UI is still using authentication_key_pairs

djberg96 replied:
Ok, I've modified `authentication_key_pairs` to return `AuthPrivateKey` and removed the separate `authentication_private_keys` method.

agrare replied:
:+1:

agrare commented on app/models/mixins/authentication_mixin.rb:
```diff
@@ -433,7 +437,7 @@ def authentication_component(type, method)
 
   def available_authentications
-    authentication_userid_passwords + authentication_key_pairs + authentication_tokens
+    authentication_userid_passwords + authentication_key_pairs + authentication_tokens + authentication_private_keys
```
Then this wouldn't be needed (key pairs == private_keys)

agrare commented on app/models/conversion_host.rb:
```diff
@@ -124,28 +208,61 @@ def resource_supports_conversion_host
+  # TODO: Move this into the provider specific subclass.
+  #
+  def find_credentials_openstack
+    AuthPrivateKey.find_by(:resource => resource) || resource.ext_management_system.authentication_type('ssh_keypair')
```
Would be better to use the `resource.authentication_*` accessors instead of doing direct find_by as this ignores the auth type.

djberg96 replied:
Ok, updated.

djberg96 commented:
@agrare I think I've addressed your concerns. The only question mark is the `available_authentications` that strangely was limited to openstack infra authkeypairs. IMO my update is more sensical.

djberg96 commented:
@agrare, just realized there's a minor issue i need to work out for cases where there's no associated authentication.

djberg96 commented:
@agrare ok, I had to update the rescue handler because otherwise we could get a false positive for `connect_ssh`. I added some more specs to cover that.

agrare commented on app/models/mixins/authentication_mixin.rb:
```diff
@@ -67,7 +67,7 @@ def authentication_tokens
 
   def authentication_key_pairs
-    authentications.select { |a| a.kind_of?(ManageIQ::Providers::Openstack::InfraManager::AuthKeyPair) }
+    authentications.select { |a| a.kind_of?(AuthPrivateKey) }
```
@djberg96 given the issues this causes on other providers I'm going to mark this refactoring hammer/no

Fryguy replied:
If I recall, this was because in public clouds we *also* store the inventory-refreshed keypairs from the providers in the authentications table, and those happen to also be AuthKeyPair.  Querying your key_pairs from the base class thus brings back both MIQ-owned and provider-owned, which is generally not wanted.

What really needs to happen is that we need to either a) move inventory-based key-pairs into a different table or b) create a separate intermediate class.

agrare replied:
@Fryguy PR added here https://github.com/ManageIQ/manageiq/pull/18633
